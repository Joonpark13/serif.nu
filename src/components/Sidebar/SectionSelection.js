import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';
import Button from '@material-ui/core/Button';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';
import Divider from '@material-ui/core/Divider';

export const styles = {
  sectionsRoot: {
    padding: 16,
  },
  title: {
    display: 'flex',
    justifyContent: 'space-between',
    alignItems: 'center',
  },
  divider: {
    marginTop: 5,
  },
  sectionTitle: {
    fontWeight: 'bold',
  },
};

function SectionSelection({
  currentCourseName,
  sections,
  back,
  classes,
}) {
  return (
    <div className={classes.sectionsRoot}>
      <Typography variant="headline" className={classes.title}>
        {currentCourseName}
        <Button onClick={back}>Back</Button>
      </Typography>
      <Divider className={classes.divider} />
      <List>
        {sections.map(section => (
          <ListItem
            key={section.id}
            button
          >
            <ListItemText>
              <Typography variant="title" className={classes.sectionTitle}>{`Section ${section.section}`}</Typography>
              <Typography>{`${section.class_mtg_info[0].meet_t}`}</Typography>
              <Typography>{`${section.class_mtg_info[0].meet_l}`}</Typography>
              <Typography>
                {section.instructor.map((teacher, idx) => (
                  `${teacher}${idx !== section.instructor.length - 1 ? ', ' : ''}`))
                }
              </Typography>
            </ListItemText>
          </ListItem>
        ))
        }
      </List>
    </div>
  );
}

SectionSelection.propTypes = {
  currentCourseName: PropTypes.string.isRequired,
  sections: PropTypes.arrayOf(PropTypes.object).isRequired,
  back: PropTypes.func.isRequired,
  classes: PropTypes.objectOf(PropTypes.string).isRequired,
};

export { SectionSelection as UnstyledSectionSelection };
export default withStyles(styles)(SectionSelection);
