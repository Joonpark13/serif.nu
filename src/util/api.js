/* eslint-disable import/prefer-default-export */
export function searchURL(query) {
  return `${process.env.API_URL}/search?term_id=4720&search_query=${query}`;
}

export function sectionsURL(termId, schoolAbbv, subjectAbbv, courseAbbv) {
  return `${process.env.API_URL}/sections?term_id=${termId}&school_abbv=${schoolAbbv}&subject_abbv=${subjectAbbv}&course_abbv=${courseAbbv}`;
}
