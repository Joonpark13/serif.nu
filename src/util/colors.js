export const northwesternPurple = '#4E2A84';
export const northwesternBrightOrange = '#FFC520';
export const northwesternRichBlack10 = '#D8D6D6';
