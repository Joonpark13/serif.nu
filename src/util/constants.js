/* eslint-disable import/prefer-default-export */
export const CURRENT_TERM = '4730';
