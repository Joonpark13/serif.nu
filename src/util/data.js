import termsData from 'data/terms.json';

/* eslint-disable import/prefer-default-export */
/* istanbul ignore next */
export const currentTerm = termsData.reduce((highestTerm, term) => {
  if (term.id > highestTerm.id) {
    return term;
  }
  return highestTerm;
});
/* eslint-enable import/prefer-default-export */
