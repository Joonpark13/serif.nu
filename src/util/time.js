export function formatMinute(minute) {
  if (minute < 10) return `0${minute}`;
  return `${minute}`;
}

export function formatTime(time, includeAmPm = true) {
  if (time.hour === 0) return `12:${formatMinute(time.minute)}${includeAmPm ? ' AM' : ''}`;
  if (time.hour < 12) return `${time.hour}:${formatMinute(time.minute)}${includeAmPm ? ' AM' : ''}`;
  if (time.hour === 12) return `12:${formatMinute(time.minute)}${includeAmPm ? ' PM' : ''}`;
  return `${time.hour - 12}:${formatMinute(time.minute)}${includeAmPm ? ' PM' : ''}`;
}

export function getFormattedClassEvent(event) {
  if (event.dow === 'TBA' || event.start === 'TBA' || event.end === 'TBA') {
    return 'TBA';
  }
  return `${formatTime(event.start, false)} - ${formatTime(event.end, false)}`;
}

export function getFormattedClassSchedule(event) {
  if (event.dow === 'TBA' || event.start === 'TBA' || event.end === 'TBA') {
    return 'TBA';
  }
  const dowStr = event.dow.join('');
  return `${dowStr} ${formatTime(event.start, true)} - ${formatTime(event.end, true)}`;
}

export function getDurationInHours(event) {
  const hoursDiff = event.end.hour - event.start.hour;
  const minutesDiff = event.end.minute - event.start.minute;
  return hoursDiff + minutesDiff / 60;
}

export function isBefore(time1, time2, orEqual = false) {
  if (time1.hour < time2.hour) {
    return true;
  }
  if (time1.hour === time2.hour) {
    if (orEqual) return time1.minute <= time2.minute;
    return time1.minute < time2.minute;
  }
  return false;
}

export function overlaps(event1, event2) {
  if (event1.dow !== event2.dow) {
    return false;
  }
  return !(isBefore(event1.end, event2.start) || isBefore(event2.end, event1.start));
}
