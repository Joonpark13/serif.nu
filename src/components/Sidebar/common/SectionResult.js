import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';
import Typography from '@material-ui/core/Typography';
import { getFormattedClassSchedule } from 'util/time';

export const styles = {
  sectionTitle: {
    fontWeight: 'bold',
  },
};

function SectionResult({ addSection, sectionHover, sectionHoverOff, section, disabled, classes }) {
  const isUnscheduled = section.schedule.some(
    schedule => schedule.dow === 'TBA' || schedule.start === 'TBA' || schedule.end === 'TBA',
  );
  return (
    <ListItem
      key={section.id}
      button
      onClick={() => addSection(section)}
      onMouseEnter={() => sectionHover(section)}
      onMouseLeave={sectionHoverOff}
      disabled={disabled || isUnscheduled}
    >
      <ListItemText>
        <Typography variant="h6" className={classes.sectionTitle}>
          {`Section ${section.sectionNumber}`}
        </Typography>

        {section.topic && <Typography>{section.topic}</Typography>}

        {section.schedule.map((scheduleObj, index) => (
          /* eslint-disable react/no-array-index-key */
          <Typography color={isUnscheduled ? 'error' : undefined} key={index}>
            {getFormattedClassSchedule(scheduleObj)}
          </Typography>
          /* eslint-enable react/no-array-index-key */
        ))}

        <Typography>{section.schedule[0].location}</Typography>

        <Typography>
          {section.instructors.map((teacher, idx) => (
            `${teacher}${idx !== section.instructors.length - 1 ? ', ' : ''}`))
          }
        </Typography>
      </ListItemText>
    </ListItem>
  );
}

SectionResult.propTypes = {
  addSection: PropTypes.func.isRequired,
  sectionHover: PropTypes.func.isRequired,
  sectionHoverOff: PropTypes.func.isRequired,
  section: PropTypes.objectOf(PropTypes.any).isRequired, // TODO
  classes: PropTypes.objectOf(PropTypes.string).isRequired,
  disabled: PropTypes.bool,
};

SectionResult.defaultProps = {
  disabled: false,
};

export { SectionResult as UnstyledSectionResult };
export default withStyles(styles)(SectionResult);
