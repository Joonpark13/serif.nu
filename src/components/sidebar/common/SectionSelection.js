import React from 'react';
import PropTypes from 'prop-types';
import { makeStyles } from '@material-ui/styles';
import { List } from '@material-ui/core';
import SectionResult from './SectionResult';
import SidebarHeader from '../SidebarHeader';

const useStyles = makeStyles({
  sectionsRoot: {
    padding: 16,
  },
  title: {
    display: 'flex',
    justifyContent: 'space-between',
    alignItems: 'center',
  },
  divider: {
    marginTop: 5,
  },
});

export default function SectionSelection({
  currentCourseName,
  sections,
  scheduledSections,
  addSection,
  back,
}) {
  const classes = useStyles();
  return (
    <div className={classes.sectionsRoot}>
      <SidebarHeader title={currentCourseName} back={back} />
      <List>
        {sections.map((section) => {
          const sectionAlreadyAdded = !!scheduledSections && scheduledSections.some(
            scheduledSection => scheduledSection.id === section.id,
          );
          return (
            <SectionResult
              key={section.id}
              section={section}
              disabled={sectionAlreadyAdded}
              addSection={addSection}
            />
          );
        })
        }
      </List>
    </div>
  );
}

SectionSelection.propTypes = {
  currentCourseName: PropTypes.string.isRequired,
  sections: PropTypes.arrayOf(PropTypes.object).isRequired,
  scheduledSections: PropTypes.arrayOf(PropTypes.object),
  addSection: PropTypes.func.isRequired,
  back: PropTypes.func.isRequired,
};

SectionSelection.defaultProps = {
  scheduledSections: [],
};
