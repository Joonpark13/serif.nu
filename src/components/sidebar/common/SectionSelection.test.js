import React from 'react';
import { shallow } from 'enzyme';
import * as timeUtils from 'util/time';
import SectionSelection from './SectionSelection';
import SectionResult from './SectionResult';

describe('SectionSelection', () => {
  const defaultProps = {
    currentCourseName: 'EECS 101-0',
    sections: [{
      id: '1',
      sectionNumber: 20,
      schedules: [{
        location: 'Some building',
      }],
      instructors: ['Jason Hartline'],
    }, {
      id: '2',
      sectionNumber: 21,
      schedules: [{
        location: 'Some other building',
      }],
      instructors: ['Ian Horswill', 'Vincent St-Amour'],
    }],
    scheduledSections: [],
    addSection: () => {},
    back: () => {},
  };
  const scheduledSectionsTestData = [{
    id: '1',
    sectionNumber: 20,
    schedules: [{
      location: 'Some building',
    }],
    instructors: ['Jason Hartline'],
  }];

  beforeEach(() => {
    timeUtils.getFormattedClassSchedule = jest.fn();
    timeUtils.getFormattedClassSchedule.mockReturnValue('schedule');
  });

  it('renders correctly', () => {
    const wrapper = shallow(<SectionSelection {...defaultProps} />);

    expect(wrapper.get(0)).toMatchSnapshot();
  });

  it('disables sections that are already scheduled', () => {
    const wrapper = shallow(
      <SectionSelection
        {...defaultProps}
        scheduledSections={scheduledSectionsTestData}
      />,
    );
    expect(
      wrapper
        .findWhere(
          section => (
            section.is(SectionResult) && section.key() === scheduledSectionsTestData[0].id
          ),
        )
        .prop('disabled'),
    ).toBe(true);
  });
});
