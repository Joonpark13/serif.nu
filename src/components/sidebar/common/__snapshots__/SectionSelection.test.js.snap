// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SectionSelection renders correctly 1`] = `
<div
  className="makeStyles-sectionsRoot-1"
>
  <SidebarHeader
    back={[Function]}
    title="EECS 101-0"
  />
  <ForwardRef(WithStyles)>
    <SectionResult
      addSection={[Function]}
      disabled={false}
      section={
        Object {
          "id": "1",
          "instructors": Array [
            "Jason Hartline",
          ],
          "schedules": Array [
            Object {
              "location": "Some building",
            },
          ],
          "sectionNumber": 20,
        }
      }
    />
    <SectionResult
      addSection={[Function]}
      disabled={false}
      section={
        Object {
          "id": "2",
          "instructors": Array [
            "Ian Horswill",
            "Vincent St-Amour",
          ],
          "schedules": Array [
            Object {
              "location": "Some other building",
            },
          ],
          "sectionNumber": 21,
        }
      }
    />
  </ForwardRef(WithStyles)>
</div>
`;
